### PR TITLE
#23 bin file upload

### PR DESCRIPTION
**Pull request by Jlopezv2050**

### CODE REVIEW

File: runtime-cli/pom.xml [hunk 1]
```diff
@@ -35,6 +35,10 @@
             <groupId>${project.groupId}</groupId>
             <artifactId>user-profile</artifactId>
         </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>file-upload</artifactId>
+        </dependency>
```
Comment (rainerkern) [inline runtime-cli/pom.xml]: Please order the dependencies alphabetically

File: pom.xml [hunk 2]
```diff
@@ -24,6 +24,7 @@
         <module>integration-tests</module>
         <module>runtime-cli</module>
         <module>user-profile</module>
+        <module>file-upload</module>
```
Comment (rainerkern) [inline pom.xml]: Please order the dependencies alphabetically

File: parent/pom.xml [hunk 3]
```diff
@@ -172,7 +182,7 @@
                 </dependencies>
             </plugin>
 
-            <!-- create test jars -->
+             <!--create test jars-->
```
Comment (rainerkern) [inline parent/pom.xml]: why?

File: parent/pom.xml [hunk 4]
```diff
@@ -147,7 +157,7 @@
         <sourceDirectory>${project.basedir}/src/main/kotlin</sourceDirectory>
         <testSourceDirectory>${project.basedir}/src/test/kotlin</testSourceDirectory>
         <plugins>
-            <!-- Kotlin compiler -->
+             <!--Kotlin compiler-->
```
Comment (rainerkern) [inline parent/pom.xml]: why?

File: parent/pom.xml [hunk 5]
```diff
@@ -97,6 +102,11 @@
                 <artifactId>runtime-cli</artifactId>
                 <version>${project.version}</version>
             </dependency>
+            <dependency>
+                <groupId>${project.parent.groupId}</groupId>
+                <artifactId>fileupload</artifactId>
+                <version>${project.version}</version>
```
Comment (rainerkern) [inline parent/pom.xml]: this is a duplicate ... I guess you wanted to add the test-jar and scope-test here

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        }
+    }
+
+    private void deleteFile(File fileToDelete){
+        if (!fileToDelete.delete()){
+            System.out.println("file has not been deleted");
+        }
+    }
+
+//                .characterEncoding("UTF-8").with(SecurityMockMvcRequestPostProcessors.csrf()).header(HttpHeaders.AUTHORIZATION,
```
Comment (rainerkern) [inline file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: please don't check in commened code

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        ResultActions resultActions = mockMvc.perform(get("/connect"))
+                .andExpect(status().isOk());
+
+        Assert.assertNotNull(resultActions.toString());
+    }
+
+    @Test
+    public void downloadFile() throws Exception {
+
+        File file = new File(fileUploadDir.concat("/downloadFile.txt"));
```
Comment (rainerkern) [inline file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: for tests please use JUnits temporary folders, 
This way they will be deleted if the test fails for any reason

https://examples.javacodegeeks.com/core-java/junit/junit-temporary-folder-example/

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+    @Autowired
+    private MockMvc mockMvc;
+
+    @Value("${file.upload-dir}")
+    private String fileUploadDir;
+
+    @Test
+    public void uploadFile() throws Exception {
+
```
Comment (rainerkern) [inline file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: please remove empty lines at the beginning of methods

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+    public void connect() throws Exception {
+        ResultActions resultActions = mockMvc.perform(get("/connect"))
+                .andExpect(status().isOk());
+
+        Assert.assertNotNull(resultActions.toString());
+    }
+
+    @Test
+    public void downloadFile() throws Exception {
+
```
Comment (rainerkern) [inline file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: please remove empty lines at the beginning of methods

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Objects;
+
+import static org.hamcrest.core.Is.is;
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.csrf;
+import static org.springframework.security.test.web.servlet.request.SecurityMockMvcRequestPostProcessors.httpBasic;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;
+import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.multipart;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;
+
+//@AutoConfigureMockMvc(secure = false)
```
Comment (rainerkern) [inline file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: please do not check in commented code

File: file-upload/src/main/kotlin/systemkern/service/FileStorageService.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+		try {
+			// Check if the file's name contains invalid characters
+			if(fileName.contains("..")) {
+				throw new FileStorageException("Sorry! Filename contains invalid path sequence " + fileName);
+			}
+
+			// Copy file to the target location (Replacing existing file with the same name)
+			Path targetLocation = this.fileStorageLocation.resolve(fileName);
+			Files.copy(file.getInputStream(), targetLocation, StandardCopyOption.REPLACE_EXISTING);
```
Comment (rainerkern) [inline file-upload/src/main/kotlin/systemkern/service/FileStorageService.java]: please add an on demand import for StandardCopyOption.REPLACE_EXISTING

File: file-upload/src/main/kotlin/systemkern/exception/MyFileNotFoundException.java [hunk 12]
```diff
@@ -0,0 +1,11 @@
+package systemkern.exception;
```
Comment (rainerkern) [inline file-upload/src/main/kotlin/systemkern/exception/MyFileNotFoundException.java]: Is this exception really necessary? ... could this not be handled with standard exceptions and HTTP error codes?

File: runtime-cli/pom.xml (hunk 1 above)
Comment (Jlopezv2050) [reply runtime-cli/pom.xml]: ✓

File: pom.xml (hunk 2 above)
Comment (Jlopezv2050) [reply pom.xml]: ✓

File: parent/pom.xml (hunk 3 above)
Comment (Jlopezv2050) [reply parent/pom.xml]: I don't know from where is this change. I recover the good one. 
rainer

File: parent/pom.xml (hunk 4 above)
Comment (Jlopezv2050) [reply parent/pom.xml]: Same than the last one.

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java (hunk 6 above)
Comment (Jlopezv2050) [reply file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: ✓

File: parent/pom.xml (hunk 5 above)
Comment (Jlopezv2050) [reply parent/pom.xml]: Yes! my fault!

File: file-upload/src/main/kotlin/systemkern/exception/MyFileNotFoundException.java (hunk 12 above)
Comment (Jlopezv2050) [reply file-upload/src/main/kotlin/systemkern/exception/MyFileNotFoundException.java]: ✓

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java (hunk 10 above)
Comment (Jlopezv2050) [reply file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: ✓

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java (hunk 8 above)
Comment (Jlopezv2050) [reply file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: ✓

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java (hunk 9 above)
Comment (Jlopezv2050) [reply file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: ✓

File: file-upload/src/main/kotlin/systemkern/service/FileStorageService.java (hunk 11 above)
Comment (Jlopezv2050) [reply file-upload/src/main/kotlin/systemkern/service/FileStorageService.java]: Is not already added?? Is not this?

rainer2

File: file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java (hunk 7 above)
Comment (Jlopezv2050) [reply file-upload/src/test/kotlin/systemkern/controller/FileControllerTest.java]: ✓

**Comment by Jlopezv2050** [comment]
Thank you for the review, I answer all your comments and I pushed the corrections. If you consider something else, I'm ready to change it  :+1:

**Pull request merged**